stdlib/runtime/numbers.gr

```diff
@@ -1721,81 +1730,286 @@ let cmpBigInt = (x: WasmI32, y: WasmI32) => {
+      },
+      t when WasmI32.eq(t, Tags._GRAIN_RATIONAL_BOXED_NUM_TAG) => {
+        let yf = WasmF64.div(
+          BI.toFloat64(boxedRationalNumerator(y)),
+          BI.toFloat64(boxedRationalDenominator(y))
+        )
+        if (WasmF64.lt(xf, yf)) -1n else if (WasmF64.gt(xf, yf)) 1n else 0n
+      },
+      t when WasmI32.eq(t, Tags._GRAIN_FLOAT32_BOXED_NUM_TAG) => {
+        let yf = WasmF64.promoteF32(boxedFloat32Number(y))
+        // special NaN cases
+        if (totalOrdering && WasmF64.ne(xf, xf)) {
+          if (WasmF64.ne(yf, yf)) {
+            0n
+          } else {
+            -1n
+          }
+        } else if (totalOrdering && WasmF64.ne(yf, yf)) {
+          if (WasmF64.ne(xf, xf)) {
+            0n
+          } else {
+            1n
+          }
+        } else {
+          if (WasmF64.lt(xf, yf)) -1n else if (WasmF64.gt(xf, yf)) 1n else 0n
+        }
+      },
+      t when WasmI32.eq(t, Tags._GRAIN_FLOAT64_BOXED_NUM_TAG) => {
+        let yf = boxedFloat64Number(y)
+        // special NaN cases
+        if (WasmF64.ne(xf, xf)) {
+          if (WasmF64.ne(yf, yf)) {
+            0n
+          } else {
+            -1n
+          }
+        } else if (WasmF64.ne(yf, yf)) {
+          if (WasmF64.ne(xf, xf)) {
+            0n
+          } else {
+            1n
+          }
+        } else {
+          if (WasmF64.lt(xf, yf)) -1n else if (WasmF64.gt(xf, yf)) 1n else 0n
+        }
+      },
+      _ => {
+        throw UnknownNumberTag
+      },
+    }
   }
 }
 
 @unsafe
-export let (>) = (x: Number, y: Number) => {
-  let xw32 = WasmI32.fromGrain(x)
-  let yw32 = WasmI32.fromGrain(y)
-  if (isBigInt(xw32)) {
-    WasmI32.gtS(cmpBigInt(xw32, yw32), 0n)
-  } else if (isBigInt(yw32)) {
-    WasmI32.leS(cmpBigInt(yw32, xw32), 0n)
+let cmpSmallInt = (x: WasmI32, y: WasmI32, is64: Bool, totalOrdering: Bool) => {
+  let xi = if (is64) {
+    boxedInt64Number(x)
   } else {
-    let xval = coerceNumberToWasmF64(x)
-    let yval = coerceNumberToWasmF64(y)
-    WasmF64.gt(xval, yval)
+    WasmI64.extendI32S(boxedInt32Number(x))
+  }
+  if (isSimpleNumber(y)) {
+    let yi = WasmI64.extendI32S(untagSimple(y))
+    if (WasmI64.ltS(xi, yi)) -1n else if (WasmI64.gtS(xi, yi)) 1n else 0n
+  } else {
+    let yBoxedNumberTag = boxedNumberTag(y)
+    match (yBoxedNumberTag) {
+      t when WasmI32.eq(t, Tags._GRAIN_INT32_BOXED_NUM_TAG) => {
+        let yi = WasmI64.extendI32S(boxedInt32Number(y))
+        if (WasmI64.ltS(xi, yi)) -1n else if (WasmI64.gtS(xi, yi)) 1n else 0n
+      },
+      t when WasmI32.eq(t, Tags._GRAIN_INT64_BOXED_NUM_TAG) => {
+        let yi = boxedInt64Number(y)
+        if (WasmI64.ltS(xi, yi)) -1n else if (WasmI64.gtS(xi, yi)) 1n else 0n
+      },
+      t when WasmI32.eq(t, Tags._GRAIN_BIGINT_BOXED_NUM_TAG) => {
+        WasmI32.sub(0n, cmpBigInt(y, x))
+      },
+      t when WasmI32.eq(t, Tags._GRAIN_RATIONAL_BOXED_NUM_TAG) => {
+        // Rationals and ints are never considered equal
```

Comment:
So `16/16 == 1` is never true? I guess it makes sense but also seems a little weird (especially if I'm adding rationals to get to `1`)

Reply:
This might need some clarification! An invariant of the number system is that rationals with a denominator of `1` do not exist, so writing the literal `16/16` gets you `1/1` which Grain just simplifies to simply `1`. So `compare(16/16, 1)` will always be zero, but this case doesn't need to worry about it because of the invariants. They're all listed in this file, but I think it might be nice to have one paragraph at the start that explains them. Shall I make an issue to get that added?

Reply:
Sure, an issue would be nice.